api/agent/agent.go

```diff
@@ -737,100 +745,155 @@ func newSizerRespWriter(max uint64, rw http.ResponseWriter) http.ResponseWriter
 
 func (s *sizerRespWriter) Write(b []byte) (int, error) { return s.w.Write(b) }
 
-func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state ContainerState) {
+// nannyHotContainer is a helper function to monitor initialization channel, container context as well as
+// agent shutdown during the lifetime of the container.
+func (a *agent) nannyHotContainer(ctx context.Context, callerCtx context.Context, cancel context.CancelFunc, evictor *EvictToken, initialized chan struct{}, pulled chan struct{}) {
+	defer cancel()
+
+	timer := time.NewTimer(a.cfg.HotPullTimeout)
+	defer timer.Stop()
+
+	// Start monitoring initialization phase. Here we expect to see "pulled" completion first,
+	// followed by "initialized".
+InitPhase:
+	for {
+		select {
+		case <-pulled: // docker-pull completed. Let's install a new timer for UDS/initialize
+			timer.Stop()
+			timer = time.NewTimer(a.cfg.HotStartTimeout)
+			pulled = make(chan struct{}) // block 'pulled' after this point
+		case <-initialized: // good, container is ready
+			break InitPhase
+		case <-ctx.Done(): // container shutdown
+			return
+		case <-a.shutWg.Closer(): // agent shutdown
+			return
+		case <-callerCtx.Done(): // original caller disconnected or serviced by another container?
+			evictor.SetEvictable(true)
+			callerCtx = context.Background() // block 'callerCtx' after this point
+		case <-evictor.C: // eviction
+			return
+		case <-timer.C: // init timeout
+			return
+		}
+	}
+
+	timer.Stop()
+
+	// Start monitoring agent shutdown
+	select {
+	case <-a.shutWg.Closer():
+	case <-ctx.Done():
+	}
```

Comment:
why do we need this goroutine to be live at this point?

Reply:
i see, because the cancel collapses the ctx done. kinda tricky, but yea makes other code cleaner

Reply:
Honestly this only saves 4-5 lines of code, and for-loop below can get away with if ctx.Err() check, but if you think this is too tricky and not worth it, I could add shutWg back to old spots.